### ISSUE TITLE
Database reached a seemingly irrecoverable state

### ISSUE THREAD

**Issue by StephanvanSchaik** [fjall-rs/fjall#68]
Hey,

Since today I am experiencing an issue with my application that uses fjall 1.1.3, in that it can no longer resume operation when starting the application and it fails on decoding one of the UTF-8 strings in the journal as shown by the backtrace below:

```
thread 'main' panicked at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/marker.rs:126:65:
should be utf-8: Utf8Error { valid_up_to: 26, error_len: Some(1) }
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
   3: core::result::Result<T,E>::expect
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1034:23
   4: <fjall::journal::marker::Marker as lsm_tree::serde::Deserializable>::deserialize
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/marker.rs:126:33
   5: <fjall::journal::reader::JournalShardReader as core::iter::traits::iterator::Iterator>::next
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/reader.rs:46:15
   6: fjall::journal::shard::JournalShard::recover_and_repair
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/shard.rs:101:25
   7: fjall::journal::Journal::recover_memtables
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/mod.rs:44:17
   8: fjall::journal::Journal::recover
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/journal/mod.rs:66:25
   9: fjall::keyspace::Keyspace::find_active_journal
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/keyspace.rs:508:32
  10: fjall::keyspace::Keyspace::recover
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/keyspace.rs:528:13
  11: fjall::keyspace::Keyspace::create_or_recover
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/keyspace.rs:279:13
  12: fjall::keyspace::Keyspace::open
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/keyspace.rs:264:24
  13: fjall::config::Config::open
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fjall-1.1.3/src/config.rs:199:9
  14: my_project::db::Database::open
             at ./src/db.rs:39:24
  15: my_project::main::{{closure}}
             at ./src/main.rs:380:14
  16: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/park.rs:282:63
  17: tokio::runtime::coop::with_budget
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/coop.rs:107:5
  18: tokio::runtime::coop::budget
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/coop.rs:73:5
  19: tokio::runtime::park::CachedParkThread::block_on
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/park.rs:282:31
  20: tokio::runtime::context::blocking::BlockingRegionGuard::block_on
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/context/blocking.rs:66:9
  21: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/scheduler/multi_thread/mod.rs:87:13
  22: tokio::runtime::context::runtime::enter_runtime
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/context/runtime.rs:65:16
  23: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
  24: tokio::runtime::runtime::Runtime::block_on
             at /home/stephan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.35.0/src/runtime/runtime.rs:350:45
  25: my_project::main
             at ./src/main.rs:522:5
  26: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Unfortunately, due to a license agreement I have for some of the data used in this database, I cannot share the files, but I would like to debug this myself and I am looking for some pointers on how to approach this since I am not familiar with the underlying disk format used by lsm-tree/fjall. Any tips would be appreciated.

**Comment by StephanvanSchaik** [comment]
I changed the code at src/journal/marker.rs:126 to show me what bytes it actually failed to parse as UTF-8, so I could then `grep` what journal files have the same byte occurrence and removing them at least causes the database to reach some state where it does not panic anymore, but I think it would be useful to figure out how it got into this state in the first place so we can prevent this from happening.

**Comment by marvin-j97** [comment]
Oh wow, thanks for looking into this. I ran workloads with >64K values before, but I think I never recovered them, so I had a false sense of security.

> are the values allowed to be longer than 64K? According to the README, the values can be up to 2^32 in length

Yes. I don't recommend large values before 2.0, but yes.

> but [the code here](https://github.com/fjall-rs/fjall/blob/main/src/journal/marker.rs#L94) seems to write out the value length as u16

Yeah that should be 32. lsm-tree supports values up to 2^32. I will add a regression test case, however changing it to u32 would be a format breaking change, so I need to look into how to handle this.

**Comment by StephanvanSchaik** [comment]
I also think that another issue is that the recovery should always reach some operational state, i.e. it should be panic-free and shouldn't use `Result<T, E>`, but it should have a way of indicating that something could be amiss with the database in these cases, such that the user can decide to turn that into a panic or simply log it somewhere. Currently, an irrecoverable journal would lead to a denial-of-service, which is not great.

> Yeah that should be 32. lsm-tree supports values up to 2^32. I will add a regression test case, however changing it to u32 would be a format breaking change, so I need to look into how to handle this.

Thanks for confirming this! I understand that it would break the format, thus that it would be incompatible with fjall 1.x. Fortunately, I can work around this for now and restructure my database again once fjall 2.x is ready.

**Comment by marvin-j97** [comment]
> I also think that another issue is that the recovery should always reach some operational state, i.e. it should be panic-free and shouldn't use Result<T, E>

I think it should be panic-free yes, however it needs to return a Result, because depending on your consistency requirements, replaying the journal may fail (kind of like https://github.com/facebook/rocksdb/wiki/WAL-Recovery-Modes). In that case, the error variant would be variant that says something like "ConsistencyFailed", and the user could decide to recover with a more lax replay mode. That's WIP however, right now everything gets recovered up until the journal is invalid (which is kTolerateCorruptedTailRecords).

> However, for fjall 1.x, it might be an idea to check that the value isn't larger than 64K when inserting data to avoid people from running into this situation. It would probably be better to have insertion fail prematurely, rather than to end up with irrecoverable journal files.

Yeah, my plan was to assert the values are not larger than 64K (they are asserted in lsm-tree to fit in a u32), and update the docs, and then v2 will support values larger than 64K, which plays well with the key-value separation feature that will come with it anyway.

**Comment by marvin-j97** [comment]
> Fortunately, I can work around this for now and restructure my database again once fjall 2.x is ready.

I have pushed a pre-release of v2 now. I don't expect any further breaking changes at this point and will probably release it on Friday (with announcement).

Here's an example for large values:

```rs
use fjall::{Config, PartitionCreateOptions};
use nanoid::nanoid;

fn main() -> fjall::Result<()> {
    let keyspace = Config::default().open()?;
    let partition = keyspace.open_partition(
        "default",
        PartitionCreateOptions::default().use_kv_separation(true),
    )?;

    let large_value = "a".repeat(128_000);
    partition.insert(nanoid!(), &large_value)?;

    for kv in partition.iter() {
        let (k, v) = kv?;
        let k = String::from_utf8_lossy(&k);
        println!("{k:?} => <{} bytes>", v.len());
    }
    println!("len: {}", partition.approximate_len());

    Ok(())
}
```

**Comment by StephanvanSchaik** [comment]
Thanks for the update! I am slowly working on migrating towards Fjall 2 and I will let you know if I run into any issues again.
